pkg/codec/modifier_codec.go

```diff
@@ -0,0 +1,185 @@
+package codec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+)
+
+func NewModifierCodec(codec types.Codec, typeProvider types.TypeProvider, modifier Modifier) (types.Codec, error) {
+	if codec == nil || typeProvider == nil || modifier == nil {
+		return nil, errors.New("inputs must not be nil")
+	}
+
+	return &modifierCodec{
+		codec:        codec,
+		typeProvider: typeProvider,
+		modifier:     modifier,
+	}, nil
+}
+
+var _ types.TypeProvider = &modifierCodec{}
+
+type modifierCodec struct {
+	codec        types.Codec
+	typeProvider types.TypeProvider
+	modifier     Modifier
+}
+
+func (m *modifierCodec) CreateType(itemType string, forEncoding bool) (any, error) {
+	t, err := m.typeProvider.CreateType(itemType, forEncoding)
+	if err != nil {
+		return nil, err
+	}
+	ot := reflect.TypeOf(t)
+	nt, err := m.modifier.RetypeForOffChain(ot)
+	if err != nil {
+		return nil, err
+	}
+
+	return reflect.Zero(nt).Interface(), nil
+}
+
+func (m *modifierCodec) Encode(ctx context.Context, item any, itemType string) ([]byte, error) {
+	tmpOnChain, err := m.typeProvider.CreateType(itemType, true)
+	if err != nil {
+		return nil, err
+	}
+
+	offChainType, err := m.modifier.RetypeForOffChain(reflect.TypeOf(tmpOnChain))
```

Comment:
`CreateType` already returns a type that went through `RetypeForOffChain`. Is it necessary to call it again? Is this done just so that we don't have to cast?

Reply:
```suggestion
	ot, err := m.CreateType(itemType, true)
	if err != nil {
		return nil, err
	}
	offChainType := reflect.TypeOf(ot)
```
Can't we do this instead? Same thing, but less err handling